src/AnalysisEpic.cxx

```diff
@@ -0,0 +1,424 @@
+      double p_ = sqrt(pow(hepmcp_psx[igen],2) + pow(hepmcp_psy[igen],2) + pow(hepmcp_psz[igen],2));
+      double mass_ = (fabs(pid_)==211)?constants::pimass:(fabs(pid_)==321)?constants::kmass:(fabs(pid_)==11)?constants::emass:(fabs(pid_)==13)?constants::mumass:(fabs(pid_)==2212)?constants::pmass:0.;
+
+      // Add to genpart
+      Particles part;
+
+      part.pid=pid_;
+      part.charge = (pid_ == 211 || pid_ == 321 || pid_ == 2212 || pid_ == -11 || pid_ == -13)?1:(pid_ == -211 || pid_ == -321 || pid_ == -2212 || pid_ == 11 || pid_ == 13)?-1:0;
+      part.mcID=igen;
+      part.vecPart.SetPxPyPzE(px_,py_,pz_,e_);
+      genpart.push_back(part);
+      
+      genidmap.insert({pz_,igen});
+
+    }
+
+    /*
+      MCParticles loop
+    */
+
+    for(int imc=0; imc < mcpart_PDG.GetSize(); imc++){
+
+      int pid_ = mcpart_PDG[imc];
+      double px_ = mcpart_psx[imc];
+      double py_ = mcpart_psy[imc];
+      double pz_ = mcpart_psz[imc];
+      double m_ = mcpart_m[imc];
+      double e_ = sqrt(px_*px_+py_*py_+pz_*pz_+m_*m_);
+
+      // Add to mcpart
+      Particles part;
+
+      part.pid=pid_;
+      part.charge = (pid_ == 211 || pid_ == 321 || pid_ == 2212 || pid_ == -11 || pid_ == -13)?1:(pid_ == -211 || pid_ == -321 || pid_ == -2212 || pid_ == 11 || pid_ == 13)?-1:0;
+      part.mcID=imc;
+      part.vecPart.SetPxPyPzE(px_,py_,pz_,e_);
+      mcpart.push_back(part);
+      
+      int igen=-1;
+      if(auto search = genidmap.find(pz_); search != genidmap.end())
+	igen=search->second; //index of the GeneratedParticle
+      
+      mcidmap.insert({imc,igen});
+
+    }
+
+    /*
+      ReconstructedParticles loop
+      - Add all particles to the std::vector<> of particles
+      - Identify the 
+      - Identify closest matching MCParticle in theta,phi,E space
+
+    */
+
+   
+      
+    for(int itrack=0; itrack < tracks_PDG.GetSize(); itrack++){
+
+      int pid_ = tracks_PDG[itrack];
+      double px_ = tracks_p_x[itrack];
+      double py_ = tracks_p_y[itrack];
+      double pz_ = tracks_p_z[itrack];
+      double e_ = tracks_e[itrack];
+      double m_ = sqrt(e_*e_-px_*px_+py_*py_+pz_*pz_);
+      
+      // Add to trackpart
+      Particles part;
+
+      part.pid=pid_;
+      part.charge = (pid_ == 211 || pid_ == 321 || pid_ == 2212 || pid_ == -11 || pid_ == -13)?1:(pid_ == -211 || pid_ == -321 || pid_ == -2212 || pid_ == 11 || pid_ == 13)?-1:0;
+      part.vecPart.SetPxPyPzE(px_,py_,pz_,e_);
+
+      /*
+	Read through Associations to match particles
+	By default, we assume no association, so mcID --> -1
+	assoc_recID --> itrack (index of the RecoParticle)
+	assoc_simID --> imc (index of the MCParticle)
+      */
+
+
+      part.mcID=-1;
+      for(int iassoc = 0 ; iassoc < assoc_simID.GetSize() ; iassoc++){
+	int idx_recID = assoc_recID[iassoc]; 
+	int idx_simID = assoc_simID[iassoc];
+	if(itrack==idx_recID){ // This track has an association
+	  part.mcID=idx_simID;
+	  break; // Only one association per particle
+	}
```

Comment:
do you find any reconstructed particles that do not have an association?

Reply:
![image](https://user-images.githubusercontent.com/27633747/205714599-987d2fc9-38b6-4b95-bd0e-33134fd3f39f.png)

This image here should showcase that indeed, there are reconstructed particles lacking an association. For each row, an event from the loaded DIS epic_arches file is shown. Since the recpartE array is not of equal size to the recID array (from associations) we infer that recID.size() <= recpartE.size()